sql/opt: add policy information to EXPLAIN for writes

spilchen opened the pull request:
This builds on the previous change that annotated verbose EXPLAIN output with policy information, which applied only to read operations. This update extends that functionality to INSERT, UPDATE, DELETE, and UPSERT operations.

Epic: CRDB-45203
Release note: none
Informs #136704

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142270)
<!-- Reviewable:end -->

michae2 reviewed (commented):
Looks great, just one question.

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142270#-:-OKgiuy91z7nrJI5ZnZT:b-npbc46)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 504 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142270#-OKgfbnP7GvkX_oyqrKN:-OKgfbnP7GvkX_oyqrKO:b-76c59q) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b83e30fae0c3fb0fa12851c731b95cda5e44968/pkg/sql/opt/exec/execbuilder/relational.go#L504)):*
> ```Go
> 			// table ID that comes from that scan.
> 			if scan, ok := e.Input.(*memo.ScanExpr); ok {
> 				annotateNodeForTable(scan.Table)
> ```

I'm a little confused by this special case for delete. Updates and upserts can also have scans as their input. Why does delete need this special case and update and upsert do not?


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142270)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 504 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142270#-OKgfbnP7GvkX_oyqrKN:-OKh8eoy7J761fgLGyQY:b-gsirmf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b83e30fae0c3fb0fa12851c731b95cda5e44968/pkg/sql/opt/exec/execbuilder/relational.go#L504)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I'm a little confused by this special case for delete. Updates and upserts can also have scans as their input. Why does delete need this special case and update and upsert do not?
</blockquote></details>

Good point. I think it might be because the queries I used in the explain test for UPDATE weren't sophisticated enough. All of the cases that I added, the input was always a `memo.ProjectExpr`. I tried many variations too. It was only the delete range case where the input was `memo.ScanExpr`. I made this logic less special and applied it to UPSERT/UPDATE. Let me know if you have any suggestions on how I can get the `memo.UpdateExpr` to be fed from a `memo.ScanExpr`.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142270)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 504 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142270#-OKgfbnP7GvkX_oyqrKN:-OKhYB8AEfTy6h5yCTjS:bg49mvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b83e30fae0c3fb0fa12851c731b95cda5e44968/pkg/sql/opt/exec/execbuilder/relational.go#L504)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Good point. I think it might be because the queries I used in the explain test for UPDATE weren't sophisticated enough. All of the cases that I added, the input was always a `memo.ProjectExpr`. I tried many variations too. It was only the delete range case where the input was `memo.ScanExpr`. I made this logic less special and applied it to UPSERT/UPDATE. Let me know if you have any suggestions on how I can get the `memo.UpdateExpr` to be fed from a `memo.ScanExpr`.
</blockquote></details>

This might get the plan you're looking for:

```
EXPLAIN (OPT, VERBOSE) UPDATE writer SET value = value WHERE key = 1;
```

But I think I'm still not understanding something. Based on your comment, why in the delete range case is the Table in the DeleteExpr not the RLS table?


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142270)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 504 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142270#-OKgfbnP7GvkX_oyqrKN:-OKkuSQR0MMfi1B-liSF:b6arb9s) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b83e30fae0c3fb0fa12851c731b95cda5e44968/pkg/sql/opt/exec/execbuilder/relational.go#L504)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

This might get the plan you're looking for:

```
EXPLAIN (OPT, VERBOSE) UPDATE writer SET value = value WHERE key = 1;
```

But I think I'm still not understanding something. Based on your comment, why in the delete range case is the Table in the DeleteExpr not the RLS table?
</blockquote></details>

The table in question is the RLS table, but multiple instances of it appear in the plan. The `Table ID` in the `DeleteExpr` differs from the one in the `ScanExpr`. Since for `DELETE` and `SELECT`, we apply filters when building the `ScanExpr`, the `Table ID` in the `ScanExpr` is the one we need to reference to correctly emit policy information.

I now have a better understanding of what’s happening. There are two types of `DELETE` operations:

1\. DELETE fed from a filtered projection

The `EXPLAIN (opt)` output for this type of deletion looks like this:
```
explain (opt) delete from t1 where value = 'stuff';
                   info
------------------------------------------
  delete t1
   └── select
        ├── scan t1
        │    ├── constraint: /7: [/1 - ]
        │    └── flags: avoid-full-scan
        └── filters
             └── value = 'stuff'
```

2\. DELETE fed directly from a scan (Delete Range)

In this case, the `EXPLAIN (opt)` output appears as follows. Here, `c1` is the primary key column:

```
explain (opt) delete from t1 where c1 = 1;
                 info
---------------------------------------
  delete t1
   └── scan t1
        ├── constraint: /7: [/1 - /1]
        └── flags: avoid-full-scan
```

For the first case, the `EXPLAIN` output for RLS policies appears when formatting the scan node. No additional changes are required:

```
explain (plan,verbose) delete from t1 where value = 'stuff';
                                           info
-------------------------------------------------------------------------------------------
  distribution: local
  vectorized: true


  • delete
  │ columns: ()
  │ estimated row count: 0 (missing stats)
  │ from: t1
  │ auto commit
  │
  └── • project
      │ columns: (c1)
      │
      └── • filter
          │ columns: (c1, value)
          │ estimated row count: 0
          │ filter: value = 'stuff'
          │
          └── • scan
                columns: (c1, value)
                estimated row count: 2 (40% of the table; stats collected 11 minutes ago)
                table: t1@t1_pkey
                spans: /1-
                policies: p1
```

For the second case, the delete range operation is fed directly by a scan. As a result, its `EXPLAIN` output for RLS policies is shown in a single node, meaning there is no opportunity to display policies for the `ScanExpr` separately, as that node isn’t explicitly emitted.

To address this, I added logic to fetch the `Table ID` from the input expression, ensuring that policy information is correctly annotated in the output. Here’s the output with my change:
```
explain (plan,verbose) delete from t1 where c1 >= -1;
                    info
--------------------------------------------
  distribution: local
  vectorized: true


  • delete range
    columns: ()
    estimated row count: 0 (missing stats)
    from: t1
    spans: /1-
    policies: p1
(9 rows)
```

This entire investigation led me to realize out that I was a bit overzealous with emitting policy information before. I even had some duplicate policy information for the UPDATE case, which has now been corrected.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> Reviewed 2 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142270)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 504 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142270#-OKgfbnP7GvkX_oyqrKN:-OKmQer37xhskStV4VZc:bcls2sf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b83e30fae0c3fb0fa12851c731b95cda5e44968/pkg/sql/opt/exec/execbuilder/relational.go#L504)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

The table in question is the RLS table, but multiple instances of it appear in the plan. The `Table ID` in the `DeleteExpr` differs from the one in the `ScanExpr`. Since for `DELETE` and `SELECT`, we apply filters when building the `ScanExpr`, the `Table ID` in the `ScanExpr` is the one we need to reference to correctly emit policy information.

I now have a better understanding of what’s happening. There are two types of `DELETE` operations:

1\. DELETE fed from a filtered projection

The `EXPLAIN (opt)` output for this type of deletion looks like this:
```
explain (opt) delete from t1 where value = 'stuff';
                   info
------------------------------------------
  delete t1
   └── select
        ├── scan t1
        │    ├── constraint: /7: [/1 - ]
        │    └── flags: avoid-full-scan
        └── filters
             └── value = 'stuff'
```

2\. DELETE fed directly from a scan (Delete Range)

In this case, the `EXPLAIN (opt)` output appears as follows. Here, `c1` is the primary key column:

```
explain (opt) delete from t1 where c1 = 1;
                 info
---------------------------------------
  delete t1
   └── scan t1
        ├── constraint: /7: [/1 - /1]
        └── flags: avoid-full-scan
```

For the first case, the `EXPLAIN` output for RLS policies appears when formatting the scan node. No additional changes are required:

```
explain (plan,verbose) delete from t1 where value = 'stuff';
                                           info
-------------------------------------------------------------------------------------------
  distribution: local
  vectorized: true


  • delete
  │ columns: ()
  │ estimated row count: 0 (missing stats)
  │ from: t1
  │ auto commit
  │
  └── • project
      │ columns: (c1)
      │
      └── • filter
          │ columns: (c1, value)
          │ estimated row count: 0
          │ filter: value = 'stuff'
          │
          └── • scan
                columns: (c1, value)
                estimated row count: 2 (40% of the table; stats collected 11 minutes ago)
                table: t1@t1_pkey
                spans: /1-
                policies: p1
```

For the second case, the delete range operation is fed directly by a scan. As a result, its `EXPLAIN` output for RLS policies is shown in a single node, meaning there is no opportunity to display policies for the `ScanExpr` separately, as that node isn’t explicitly emitted.

To address this, I added logic to fetch the `Table ID` from the input expression, ensuring that policy information is correctly annotated in the output. Here’s the output with my change:
```
explain (plan,verbose) delete from t1 where c1 >= -1;
                    info
--------------------------------------------
  distribution: local
  vectorized: true


  • delete range
    columns: ()
    estimated row count: 0 (missing stats)
    from: t1
    spans: /1-
    policies: p1
(9 rows)
```

This entire investigation led me to realize out that I was a bit overzealous with emitting policy information before. I even had some duplicate policy information for the UPDATE case, which has now been corrected.
</blockquote></details>

Ah, thanks for the explanation. I didn't realized DeleteRange was built out of a normal DeleteExpr on top of a ScanExpr.

This makes sense, then. (If you wanted to, you might be able to factor out the [check](https://github.com/cockroachdb/cockroach/blob/a6b90eb338acbd09c540a1c47c067fa9d05cd175/pkg/sql/opt/exec/execbuilder/mutation.go#L651-L678) at the top of tryBuildDeleteRange into a helper function, to get a more accurate condition. But this seems fine, too.)

(The other interesting operator like this is InsertFastPath, but I'm guessing you won't need to do anything special for that, since it doesn't incorporate a read of the same table.)
___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 497 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/142270#-OKmStSl0KE6E-YZNREQ:-OKmStSl0KE6E-YZNRER:b5qxmv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2df4e2babc350f0cdec67b9298e82962fd5d20d/pkg/sql/opt/exec/execbuilder/relational.go#L497)):*
> ```Go
> 			annotateNodeForTable(e.Table)
> 		case *memo.InsertExpr:
> 			annotateNodeForTable(e.Table)
> ```

I might be misunderstanding something about RLS, but if Insert is annotated, then I would think Update and Upsert would also need to be, since those operators can also "create new rows" out of thin air.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142270)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 504 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142270#-OKgfbnP7GvkX_oyqrKN:-OKzy1IXDXVVUFmvGSg5:b-12oytv) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b83e30fae0c3fb0fa12851c731b95cda5e44968/pkg/sql/opt/exec/execbuilder/relational.go#L504)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Ah, thanks for the explanation. I didn't realized DeleteRange was built out of a normal DeleteExpr on top of a ScanExpr.

This makes sense, then. (If you wanted to, you might be able to factor out the [check](https://github.com/cockroachdb/cockroach/blob/a6b90eb338acbd09c540a1c47c067fa9d05cd175/pkg/sql/opt/exec/execbuilder/mutation.go#L651-L678) at the top of tryBuildDeleteRange into a helper function, to get a more accurate condition. But this seems fine, too.)

(The other interesting operator like this is InsertFastPath, but I'm guessing you won't need to do anything special for that, since it doesn't incorporate a read of the same table.)
</blockquote></details>

​I like your suggestion to add a helper function; it clarifies the code that checks for delete range in `maybeAnnotatePolicyInfo`.

Regarding `InsertFastPath`, we've already addressed it by handling `InsertExpr` within `maybeAnnotatePolicyInfo` and invoking `emitPolicies` for `insertFastPathOp`. However, I realized that my EXPLAIN test didn't cover that op, so I've updated it accordingly.
___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 497 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/142270#-OKmStSl0KE6E-YZNREQ:-OL-8GLz8aF6IE-TMl-O:b-b0ra8h) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2df4e2babc350f0cdec67b9298e82962fd5d20d/pkg/sql/opt/exec/execbuilder/relational.go#L497)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I might be misunderstanding something about RLS, but if Insert is annotated, then I would think Update and Upsert would also need to be, since those operators can also "create new rows" out of thin air.
</blockquote></details>

I initially omitted the `annotateNodeForTable` calls for `UPDATE` and `UPSERT` operations because both are driven by a scan, and the RLS information is displayed at the scan node. 

However, I realized that we apply policies differently for new and existing rows. For instance, in an `UPDATE`, distinct policies can govern the reading of existing rows versus the validation of new rows. Therefore, it is probably a better design to track how these policies are applied.

So, I have reinstated the `annotateNodeForTable` calls for both `UPDATE` and `UPSERT`. And the policy tracking now differentiates between policies used to filter existing rows and those applied as check constraints for new rows.


<!-- Sent from Reviewable.io -->

spilchen commented:
PTAL @michae2

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Very nice! Thanks for those changes.

> Reviewed 7 of 7 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142270#-:-OL6aTon6qhJu1gCuytY:b8gb05)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/opt/exec/explain/testdata/row_level_security` line 257 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/142270#-OL6_9_PBoQ3iuKOeyqw:-OL6_9_PBoQ3iuKOeyqx:b-ga3jll) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa0569bf7e50488c001648f96af01144806b5be2/pkg/sql/opt/exec/explain/testdata/row_level_security#L257)):*
> ```
> 
> exec-ddl
> CREATE POLICY p_update ON writer FOR UPDATE USING (key = 5)  WITH CHECK (value like 'updater: %')
> ```

It might be nice to see two different policies in the testcase EXPLAIN output, one for reading and one for writing. Maybe that would be something like the following (not sure if I have the syntax right)?

```sql
CREATE POLICY p_update_read ON writer FOR UPDATE USING (key = 7);
CREATE POLICY p_update_write ON writer FOR UPDATE WITH CHECK (value like 'updater: %');
```


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/opt: add policy information to EXPLAIN for writes

This builds on the previous change that annotated verbose EXPLAIN output
with policy information, which applied only to read operations. This
update extends that functionality to INSERT, UPDATE, DELETE, and UPSERT
operations.

Epic: CRDB-45203
Release note: none
Informs #136704

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142270)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2)

___
*[`pkg/sql/opt/exec/explain/testdata/row_level_security` line 257 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/142270#-OL6_9_PBoQ3iuKOeyqw:-OL95IlY26GFv2b-deu-:bma3e2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa0569bf7e50488c001648f96af01144806b5be2/pkg/sql/opt/exec/explain/testdata/row_level_security#L257)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

It might be nice to see two different policies in the testcase EXPLAIN output, one for reading and one for writing. Maybe that would be something like the following (not sure if I have the syntax right)?

```sql
CREATE POLICY p_update_read ON writer FOR UPDATE USING (key = 7);
CREATE POLICY p_update_write ON writer FOR UPDATE WITH CHECK (value like 'updater: %');
```
</blockquote></details>

This is a good idea. Unfortunately, we don’t yet have everything in place to test multiple policies on a table. That functionality is coming in #142479. In the meantime, I’ve updated the test so we can verify this once support for multiple policies is added.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
